[D2] 모든 Button Component UI 구현

kimkyunghun3 opened the pull request:
# 개요
- 🔗  이슈링크 : #5 


# 변경사항

## 작업 내용
- CategoryButton(이전에 PR 올렸던 버튼 코드 수정)
가변버튼
사용처에서 입력할 필요 없는 코드를 Default 값 주어지고 입력하지 않도록 구현
```swift
self.layer.cornerRadius = 20
self.titleLabel?.font = .systemFont(ofSize: 14)
```

- CardListButton(List Card에 쓰이는 버튼)
고정길이 버튼
고정 width를 가지고 내부에 필요한 title를 enum으로 구성하여 선택하도록 구현
```swift
// CardListButtonStyle.swift 
public enum TitleStatus: String {
    case apply = "응모하기"
    case complete = "응모완료"
    case result = "결과확인"
    case progress = "응모중"
    case failure = "꽝"
    case win = "당첨"
  }

// CardListButton.swift
NSLayoutConstraint.activate([
    self.heightAnchor.constraint(equalToConstant: 34),
    self.widthAnchor.constraint(equalToConstant: 96)
])
```

- AlarmButton(알림 메시지 및 당첨결과 페이지의 결과확인, 확인완료 버튼)
가변 버튼
title enum으로 구성
```swift
public enum TitleStatus: String {
    case checkResult = "결과 확인"
    case confirmed = "확인 완료"
}
```

- CTAButton(양옆을 꽉채우는 CTA 모든 Button)
사용처에서 leading, trailing를 사용하여 양옆을 꽉채우는 버튼

### 고민 포인트
CategoryButton, AlarmButton 두 버튼이 가변 버튼인데, radius, inset 등을 사용처에서 준다면 하나의 버튼으로 합칠 수도 있을 것 같다. 하지만 이는 사용처에서 많은 입력을 요구하고 각 inset, radius를 일일이 확인해야하는 불편함이 존재할 것 같다. 또한 title 부분이 불필요하게 양쪽에서 입력이 주어질 것같다. 그러므로 분리를 통해 사용처에서 이를 입력하지 않도록 구현했는데 괜찮은 방향일지 궁금하다.


### 스크린샷
**- CategoryButton**

| 가변 버튼 | 가변 버튼
|:---:|:---:|
|<img src ="https://user-images.githubusercontent.com/52434820/177897749-2ecc58ac-e0e0-4e9f-bed5-23d554f5a606.png" width ="400">|<img src ="https://user-images.githubusercontent.com/52434820/177897757-21353246-28ad-403c-bea9-f3c01fd2ceba.png" width ="400">|


**- CardListButton**

| 고정 버튼 | 고정 버튼
|:---:|:---:|
|<img src ="https://user-images.githubusercontent.com/52434820/177897801-8f33665a-5680-4a9e-9a51-5df8f0fd02e7.png" width = "400">|<img src ="https://user-images.githubusercontent.com/52434820/177897869-ec21c752-dd3a-4605-a33b-8375545d5d38.png" width ="400">|

**- AlarmButton**

| 가변 버튼 |
|:---:|
|<img src ="https://user-images.githubusercontent.com/52434820/177897782-46e2c0a5-20cf-49fc-a799-b166e4a2228f.png" width ="400">|

**- CTAButton**

| 12 Max | SE |
|:---:|:---:|
|<img src ="https://user-images.githubusercontent.com/52434820/177896813-ceb9cb50-634d-4969-85eb-d88daf5f0b3d.png" width ="400">|<img src ="https://user-images.githubusercontent.com/52434820/177896836-72426a9e-e740-4bad-b67f-e7cf164a3dfb.png" width ="400">|

kimkyunghun3 committed:
:recycle: DDIPButton 변하지않는 값 기본값 부여 #5

# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

################
# 세부 내용을 아랫줄에 작성

################
# 추가적인 세부 내용을 아랫줄에 작성

################
# :tada: = 프로젝트 시작
# :sparkles: = create 파일 생성 시
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시
# :rocket: = release 배포 시
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우)
# :memo: = 문서 작성 시
# :art: = 디자인 시스템 관련 내용 (Color / Asset)
# :hammer: = 테스트 코드 작성
################

kimkyunghun3 committed:
:bulb: 응모박스 CardListButtonStyle 구현 #5

# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

################
# 세부 내용을 아랫줄에 작성

################
# 추가적인 세부 내용을 아랫줄에 작성

################
# :tada: = 프로젝트 시작
# :sparkles: = create 파일 생성 시
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시
# :rocket: = release 배포 시
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우)
# :memo: = 문서 작성 시
# :art: = 디자인 시스템 관련 내용 (Color / Asset)
# :hammer: = 테스트 코드 작성
################

kimkyunghun3 committed:
:bulb: 응모박스 CardListButton 구현 #5

# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

################
# 세부 내용을 아랫줄에 작성

################
# 추가적인 세부 내용을 아랫줄에 작성

################
# :tada: = 프로젝트 시작
# :sparkles: = create 파일 생성 시
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시
# :rocket: = release 배포 시
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우)
# :memo: = 문서 작성 시
# :art: = 디자인 시스템 관련 내용 (Color / Asset)
# :hammer: = 테스트 코드 작성
################

kimkyunghun3 committed:
:bulb: 알람 ButtonStyle 구현 #5

# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

################
# 세부 내용을 아랫줄에 작성

################
# 추가적인 세부 내용을 아랫줄에 작성

################
# :tada: = 프로젝트 시작
# :sparkles: = create 파일 생성 시
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시
# :rocket: = release 배포 시
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우)
# :memo: = 문서 작성 시
# :art: = 디자인 시스템 관련 내용 (Color / Asset)
# :hammer: = 테스트 코드 작성
################

kimkyunghun3 committed:
:bulb: 알람 버튼 구현 #5

# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

################
# 세부 내용을 아랫줄에 작성

################
# 추가적인 세부 내용을 아랫줄에 작성

################
# :tada: = 프로젝트 시작
# :sparkles: = create 파일 생성 시
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시
# :rocket: = release 배포 시
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우)
# :memo: = 문서 작성 시
# :art: = 디자인 시스템 관련 내용 (Color / Asset)
# :hammer: = 테스트 코드 작성
################

kimkyunghun3 committed:
:bulb: CTA 버튼 스타일 구현 #5

# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

################
# 세부 내용을 아랫줄에 작성

################
# 추가적인 세부 내용을 아랫줄에 작성

################
# :tada: = 프로젝트 시작
# :sparkles: = create 파일 생성 시
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시
# :rocket: = release 배포 시
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우)
# :memo: = 문서 작성 시
# :art: = 디자인 시스템 관련 내용 (Color / Asset)
# :hammer: = 테스트 코드 작성
################

kimkyunghun3 committed:
:bulb: CTA 버튼 구현 #5

# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

################
# 세부 내용을 아랫줄에 작성

################
# 추가적인 세부 내용을 아랫줄에 작성

################
# :tada: = 프로젝트 시작
# :sparkles: = create 파일 생성 시
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시
# :rocket: = release 배포 시
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우)
# :memo: = 문서 작성 시
# :art: = 디자인 시스템 관련 내용 (Color / Asset)
# :hammer: = 테스트 코드 작성
################

kimkyunghun3 committed:
:recycle: Button layout 에러 수정 #5

# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

################
# 세부 내용을 아랫줄에 작성

################
# 추가적인 세부 내용을 아랫줄에 작성

################
# :tada: = 프로젝트 시작
# :sparkles: = create 파일 생성 시
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시
# :rocket: = release 배포 시
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우)
# :memo: = 문서 작성 시
# :art: = 디자인 시스템 관련 내용 (Color / Asset)
# :hammer: = 테스트 코드 작성
################

kimkyunghun3 committed:
:recycle: Button -> CategoryButton 네이밍 수정 #5

# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

################
# 세부 내용을 아랫줄에 작성

################
# 추가적인 세부 내용을 아랫줄에 작성

################
# :tada: = 프로젝트 시작
# :sparkles: = create 파일 생성 시
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시
# :rocket: = release 배포 시
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우)
# :memo: = 문서 작성 시
# :art: = 디자인 시스템 관련 내용 (Color / Asset)
# :hammer: = 테스트 코드 작성
################

AhnSangHee reviewed (approved):
고생하셨습니다~!

szzang-kr commented on Projects/DesignSystem/Sources/Component/Button/DDIPButton.swift:
```diff
@@ -23,12 +22,15 @@ public class DDIPButton: UIButton {
     public func setUI() {
-        self.layer.cornerRadius = self.style.radius
+        self.layer.cornerRadius = 20
         self.setTitleColor(self.style.titleColor, for: .normal)
         self.backgroundColor = self.style.buttonColor
-        self.titleLabel?.font = self.style.buttonFont
+        self.titleLabel?.font = .systemFont(ofSize: 14)
         self.setTitle(self.style.title, for: .normal)
         self.contentEdgeInsets = UIEdgeInsets(top: self.style.topInset, left: self.style.leftInset, bottom: self.style.bottomInset, right: self.style.rightInset)
```
👍

szzang-kr commented on Projects/DesignSystem/Sources/Component/Button/DDIPAlarmButton.swift:
```diff
@@ -0,0 +1,39 @@
+        self.backgroundColor = self.style.buttonColor
+        self.titleLabel?.font = .systemFont(ofSize: 12)
+        self.setTitle(self.style.title?.rawValue, for: .normal)
+        self.contentEdgeInsets = UIEdgeInsets(top: self.style.topInset, left: self.style.leftInset, bottom: self.style.bottomInset, right: self.style.rightInset)
```
한줄로 너무 길게 표현되는 코드는 다음과 같이 변경하면 한눈에 알아보기 좋아요!

다만 요것은 컨벤션이니까 이렇게 하는 방법도 있구나 라고 알아만 두셔도 좋아요 😄 

```swift 
self.contentEdgeInsets = UIEdgeInsets(
    top: self.style.topInset,
    left: self.style.leftInset,
    bottom: self.style.bottomInset,
    right: self.style.rightInset
)
```

kimkyunghun3 replied:
말씀해주신 컨벤션이 가독성이 높아지는 것 같네요!
수정했습니다 감사합니다 :)

szzang-kr reviewed (approved):
고생하셨어요 ~~ ⭐

kimkyunghun3 committed:
:recycle: alarmButton 코드 컨벤션 변경 #5

# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

################
# 세부 내용을 아랫줄에 작성

################
# 추가적인 세부 내용을 아랫줄에 작성

################
# :tada: = 프로젝트 시작
# :sparkles: = create 파일 생성 시
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시
# :rocket: = release 배포 시
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우)
# :memo: = 문서 작성 시
# :art: = 디자인 시스템 관련 내용 (Color / Asset)
# :hammer: = 테스트 코드 작성
################

kimkyunghun3 committed:
:recycle: radius 동그랗게 나오도록 수정 #5

# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

################
# 세부 내용을 아랫줄에 작성

################
# 추가적인 세부 내용을 아랫줄에 작성

################
# :tada: = 프로젝트 시작
# :sparkles: = create 파일 생성 시
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시
# :rocket: = release 배포 시
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우)
# :memo: = 문서 작성 시
# :art: = 디자인 시스템 관련 내용 (Color / Asset)
# :hammer: = 테스트 코드 작성
################
